### ISSUE TITLE
See if we can somehow use .gitignore

### ISSUE THREAD

**Issue by petdance** [petdance/ack3#40]
_From @petdance on March 18, 2012 2:10_

I don't think we want to use .gitignore automatically, but maybe in ack 2.1 we can do something like add a --ignorefile=.gitignore.

_Copied from original issue: petdance/ack2#28_

**Comment by petdance** [comment]
_From @dnblankedelman on March 18, 2012 2:18_

And perhaps support .hgignore as well? I'm not an hg user, but I bet others would find that helpful as well.

**Comment by petdance** [comment]
That's why I figured it would be a --ignorefile and people could specify whatever file they wanted.

Although now I see that .gitignore and .hgignore use different formats.

**Comment by petdance** [comment]
_From @hoelzro on March 18, 2012 11:57_

What do you think of doing plugins for Ack 2.1?  I think this would be a great example plugin.

**Comment by petdance** [comment]
Yes, I think we should do plugins for 2.1

This feature, however, I don't think should need to be a plugin.  I think it would be simple enough to do without having to have a plugin.  The big point of plugins is, in my mind, to include code that would take up a lot of code or execution time in the base ack, or would require a module.  I don't think this feature would do either of those things.

**Comment by petdance** [comment]
_From @hoelzro on March 18, 2012 14:15_

Hmm...I suppose we differ on what should and shouldn't be a plugin.  I think a generic ignore file is "core-worthy", but as soon as we start to cater to specific use case (ex. Git/Hg integration), that's what I consider "plugin-worthy".

**Comment by petdance** [comment]
Actually, I'd make .gitignore standard, and .hgignore a plugin because it's more complex.

But this is all down the road.

**Comment by petdance** [comment]
_From @hoelzro on May 11, 2013 16:24_

See also #144, #272.

**Comment by petdance** [comment]
_From @georgexsh on February 12, 2014 6:23_

+1 for this
with git, ignore file not only in `.gitignore`, but also `$GIT_DIR/info/exclude`, see gitignore(5)

**Comment by petdance** [comment]
_From @moll on May 22, 2014 15:59_

Any news on this? Any plugins or anything?

**Comment by petdance** [comment]
Sorry, no.  We don't even have a plugin architecture yet. :-(

**Comment by petdance** [comment]
_From @medikoo on October 24, 2014 9:18_

+1

Following what's in `.gitignore` is mostly what I expect, and current defaults are too invasive

**Comment by petdance** [comment]
_From @sheerun on January 20, 2015 1:26_

+1 `--ignore-dir` entires in ackrc grow larger and larger, because I can't simply ignore what's in gitignore

**Comment by petdance** [comment]
_From @hoelzro on January 20, 2015 4:41_

Thanks everyone for chiming in.  ack2 is in a bit of a code freeze at the moment; @petdance would like to go through everything that's changed since 2.14 so that 2.16 can be released, and we don't want to heap any more work on at the moment.  If anyone is willing to submit a pull request implementing the feature, however, we can merge it in once 2.16 is released!

**Comment by petdance** [comment]
.gitignore is all globs, right?  The .gitignore would be the only place that ack handles globs.

Does anyone know how Windows Git does .gitignore?

Handling .gitignore is not a CYJ.  (Can't you just...)

**Comment by petdance** [comment]
_From @hoelzro on January 20, 2015 17:38_

I don't know for sure, but I'm guessing that gitignore on Windows works pretty much the same as on OS X/Linux, maybe with some accounting for case insensitivity.

.gitignore can contain more than just globs; you can have inverted expressions (ex. ignore all `.jar` files, except for those under `deps`), or relative paths (ex. ignore all jars in `/.mvn` or something).  Absolute paths introduce an especially sinister level of complexity, since they're absolute with respect to the root of the working tree.  Which means we would have to detect the root of the working tree, which means respecting `GIT_WORK_TREE` and `GIT_DIR`.  And let's not forget that there's not just one gitignore file; you have `.gitignore`, `$GIT_DIR/info/exclude`, the config setting `core.excludesfile`, and anything else I forgot to mention.  Long story short, that would be a _lot_ of git implementation details to put into ack, and frozen to which ever version's implementation we would port over.

That being said, we could shell out to `git ls-files`, but then we have to handle the case where someone doesn't have git installed, or isn't in a git repository, and reconcile potential differences between versions of Git.  So even that route is frought with peril.

My suggestion is this:  if you really want this feature, write a shell wrapper around `ack` to call `git ls-files` and populate `ACK_OPTIONS` based on its output, and then call `ack`.  If we end up implementing this feature, it probably won't be for a while.

**Comment by petdance** [comment]
_From @SethTisue on January 20, 2015 18:54_

fwiw, from issues like e.g. https://github.com/ggreer/the_silver_searcher/issues/100 it appears the ag folks went the route of parsing `.gitignore` themselves (apparently imperfectly)

**Comment by petdance** [comment]
_From @pmorch on March 12, 2016 12:42_

Perhaps one way to achieve this without supporting `.gitigore` directly in `ack` could be to write a script that generates a `$PROJECT/.ackrc` from all the `.gitignore` files (and whatnot) in the project.

Probably not trivial to write if all the corner cases must be supported. @hoelzro [wrote](https://github.com/petdance/ack2/issues/28#issuecomment-70696864):

> And let's not forget that there's not just one gitignore file; you have .gitignore, $GIT_DIR/info/exclude, the config setting core.excludesfile, and anything else I forgot to mention.

This would certainly scratch my itch right now:

```
cd $PROJECT
$ACK/contrib/generate-ackrc-from-gitignore.pl . > .ackrc
ack pattern
```

Also, then it isn't baked directly into `ack`.

**Comment by petdance** [comment]
_From @naggie on February 1, 2017 12:28_

`alias gack="git ls-files -oc --exclude-standard | ack -x"` is what I use.

**Comment by petdance** [comment]
Does my heart good to see people using `ack -x`.

**Comment by petdance** [comment]
I don't want ack to have to dig into .gitignore files, but we can put the git line in the docs.

**Comment by petdance** [comment]
_From @naggie on March 16, 2017 9:45_

Good idea, I agree with you that ack should not parse .gitignore files.

One caveat with that alias is that submodules are listed by the git command and they are directories -- ack then (correctly) complains about the directories.


With git 2.11+, the `--recurse-submodules` option may be used which should work around this, and of course match files within submodules which might not be desired.


Here's the same command with the long options which are semantic so better for docs:
```
alias gack="git ls-files --others --cached --exclude-standard | ack -x"
```

**Comment by petdance** [comment]
We are not going to parse .gitignore.

**Comment by n1vux** [comment]
re parsing .gitignore, which above comment says we aren't doing, there's a new modules or two that would outsource doing that if we don't want to do it ourselves but would like to be able to DWIM.

> Text::Gitignore - Match .gitignore patterns
> It combines Text::Glob and File::FnMatch functionality with several ".gitignore"-specific tweaks.... 

> Parse::Gitignore - parse .gitignore files [does matching too]

Parse a .gitignore file and check whether a file matches it....
 https://metacpan.org/search?q=gitignore

**Comment by petdance** [comment]
No, if you give ack a list of files you're going to process, then it ignores all type filtering.  It's the same as if you typed `ack --js foo.php`, you'd be searching the `foo.php` file, despite the `--js`.  If you specify a list of files, ack assumes you know that's what you want.
